pkg/metricfamily/whitelist.go

```diff
@@ -0,0 +1,63 @@
+package metricfamily
+
+import (
+	clientmodel "github.com/prometheus/client_model/go"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+)
+
+type whitelist [][]*labels.Matcher
+
+// NewWhitelist returns a Transformer that checks if at least one
+// rule in the whitelist is true.
+// This Transformer will nil metrics within a metric family that do not match a rule.
+// Each given rule is transformed into a matchset. Matchsets are OR-ed.
+// Individual matchers within a matchset are AND-ed, as in PromQL.
+func NewWhitelist(rules []string) (Transformer, error) {
+	var ms [][]*labels.Matcher
+	for i := range rules {
+		matchers, err := promql.ParseMetricSelector(rules[i])
+		if err != nil {
+			return nil, err
+		}
+		ms = append(ms, matchers)
+	}
+	return whitelist(ms), nil
+}
+
+// Transform implements the Transformer interface.
+func (t whitelist) Transform(family *clientmodel.MetricFamily) (bool, error) {
+	var keep bool
+Metric:
+	for i, m := range family.Metric {
+		if m == nil {
+			continue
+		}
+		for _, matchset := range t {
+			if match(family.GetName(), m, matchset...) {
+				keep = true
+				continue Metric
+			}
+		}
+		family.Metric[i] = nil
+	}
+	return keep, nil
+}
+
+// match checks whether every Matcher matches a given metric.
+func match(name string, metric *clientmodel.Metric, matchers ...*labels.Matcher) bool {
```

Comment:
@brancz is there a more semantic way to implement this match check against a family using a Prometheus library? The only other _nice_ option I though of was loading the metric into a temporary tsdb storage struct so I could use the storage.Querier interface, however that seemed a little heavy duty for this simple check.

Reply:
That is definitely too heavy, also it wouldn’t really satisfy the interface we’re looking for here. I would just suggest a function that takes a labelset and a slice of matchers, and each label of the labelset must pass every marcher. If we make sure the matchers (including Regex matchers) are compiles at start time of telemeter we should be good.

Reply:
The underlying type of a `LabelSet` is a `map[LabelName]LabelValue`:

https://github.com/prometheus/common/blob/4724e9255275ce38f7179b2478abeae4e28c904f/model/labelset.go#L28

I do agree that the signature would look a little cleaner, but the current code is a little bit more efficient. The current code also already ensures compilation in the constructor.